Streamline a few members of char

stephentoub opened the pull request:
- Remove a branch from IsLetterOrDigit
- Slightly shrink the asm for IsPunctuation and IsSymbol
- Remove a bounds check from IsSurrogatePair
- Remove some bounds checks, other branching, and duplicative code from ConvertToUtf32

Also replaced a few uses of CharUnicodeInfo.GetUnicodeCategory with char.GetUnicodeCategory where ASCII chars were likely inputs.

|             Method |         Toolchain |     Mean |   Error |  StdDev | Ratio | Code Size |
|------------------- |------------------ |---------:|--------:|--------:|------:|----------:|
| CountLetterOrDigit | \main\corerun.exe | 262.5 us | 2.42 us | 2.26 us |  1.00 |     162 B |
| CountLetterOrDigit |   \pr\corerun.exe | 232.0 us | 4.48 us | 4.98 us |  0.88 |     133 B |
|                    |                   |          |         |         |       |           |
|        CountSymbol | \main\corerun.exe | 260.6 us | 1.06 us | 0.94 us |  1.00 |     132 B |
|        CountSymbol |   \pr\corerun.exe | 244.7 us | 2.29 us | 1.79 us |  0.94 |     124 B |
|                    |                   |          |         |         |       |           |
|   CountPunctuation | \main\corerun.exe | 261.4 us | 0.88 us | 0.78 us |  1.00 |     132 B |
|   CountPunctuation |   \pr\corerun.exe | 244.8 us | 2.49 us | 2.08 us |  0.94 |     124 B |
|                    |                   |          |         |         |       |           |
|     ConvertToUtf32 | \main\corerun.exe | 309.7 us | 5.71 us | 5.34 us |  1.00 |     583 B |
|     ConvertToUtf32 |   \pr\corerun.exe | 243.6 us | 3.62 us | 3.21 us |  0.79 |     330 B |

```C#
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Diagnosers;
using BenchmarkDotNet.Running;
using System.Linq;

[DisassemblyDiagnoser]
public partial class Program
{
    static void Main(string[] args) => BenchmarkSwitcher.FromAssembly(typeof(Program).Assembly).Run(args);

    private string _input = string.Concat(Enumerable.Range(0, char.MaxValue + 1).Select(i => (char)i));
    private string _smiles = string.Concat(Enumerable.Repeat("\ud83d\ude00", 100_000));

    [Benchmark]
    public int CountLetterOrDigit()
    {
        int total = 0;
        foreach (char c in _input) if (char.IsLetterOrDigit(c)) total++;
        return total;
    }

    [Benchmark]
    public int CountSymbol()
    {
        int total = 0;
        foreach (char c in _input) if (char.IsSymbol(c)) total++;
        return total;
    }

    [Benchmark]
    public int CountPunctuation()
    {
        int total = 0;
        foreach (char c in _input) if (char.IsPunctuation(c)) total++;
        return total;
    }

    [Benchmark]
    public int ConvertToUtf32()
    {
        int total = 0;
        string smiles = _smiles;
        for (int i = 0; i < smiles.Length; i += 2) total += char.ConvertToUtf32(smiles, i);
        return total;
    }
}
```

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Char.cs:
```diff
@@ -850,12 +846,9 @@ public static bool IsSymbol(string s, int index)
-            }
-
-            return CheckSymbol(CharUnicodeInfo.GetUnicodeCategoryInternal(s, index));
+            return CheckSymbol(IsLatin1(c) ?
+                GetLatin1UnicodeCategory(c) :
+                CharUnicodeInfo.GetUnicodeCategoryInternal(s, index));
```
Wonder if this is a transform we could have the JIT implicitly do...

Would require some pattern recognition in morph, but seems "feasible" overall?

stephentoub replied:
(That said, it's possible this change would actually be a deoptimization if everything gets inlined and `c` is a known constant?)

stephentoub replied:
> I could definitely see it being a de-optimization in some cases, but ideally the JIT could have some basic heuristics and constant folding for this to avoid most of the issues.

When I said "this", I was actually referring to my change :)

tannergooding replied:
Wouldn't expect this change to impact existing constant folding support and I'd expect it to improve inlinability in general.

A local check shows both get optimized down correctly for `IsPunctuation(' ')`:
```asm
; Method T:B():bool
G_M000_IG01:                ;; offset=0000H
       4883EC28             sub      rsp, 40

G_M000_IG02:                ;; offset=0004H
       B920000000           mov      ecx, 32
       FF15F9792E00         call     [GetLatin1UnicodeCategory(ushort):int]
       8BC8                 mov      ecx, eax
       FF15C1792E00         call     [CheckPunctuation(int):bool]
       90                   nop      

G_M000_IG03:                ;; offset=0018H
       4883C428             add      rsp, 40
       C3                   ret      
; Total bytes of code: 29
```

tannergooding replied:
The real method fully optimizes both to `xor eax, eax`

stephentoub replied:
> The real method fully optimizes both to xor eax, eax

:smile:

tannergooding commented on src/libraries/System.Private.CoreLib/src/System/Char.cs:
```diff
@@ -1000,7 +993,7 @@ public static bool IsSurrogatePair(string s, int index)
             }
 
-            if (index + 1 < s.Length)
+            if ((uint)(index + 1) < (uint)s.Length)
```
Why is this cast needed, the above check should have ensured `index` is non-negative and the length of a string can't be `int.MaxValue`?

stephentoub replied:
It may not be with recent improvements.  It was on .NET 7, according to [SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgBEBLbAcwDsJcGDmFwsAwhAAmMAIK9sAGwCeuDrgDcNGsQDMTUgzEMA3jQbmme5kgbAIEBQwCSuAMoBXKFAjdsGGAAK2BxQ5AAUzIy4aAwcvBixvNIIAJRmFqbUFtmxAGYMYbgMALzFDLzuCgppWTnmmXV1ACoAFt4A7mEpmrV1AL7pdRz5YWHucRgpcckMAHxlYxMpogAyMLzcGC01jSaDuwytHV09uwPU+9nDBdMwCAwA1AzkDAA8DKvrm9uXGb91xAA7AwwC1sFAWC4PF4fH5AsEoIUANq3BAAXRiuBRSTuj2eaO6/wY512QIYuUUuBgpwsJIs+10THINjsDmcbk83l8/iCIVIEXIURiE0SyR2f16OWuhRKZQqVXFdQaB3MRwgnUJkuydKGI0W8SmOPu8wK4wNnw2W0VOWVKrVGpp/S0Wos0v1kzCqLx5BSb1NSwt32t2VtBzJoPBkI5MO58JCyNRGI+2JmT3IBMdOR1OTJFIUVMz2f2SLYEBguAAchAMAAlGAYTy8NH7AAOUA4ADc4UybMQUIc2uquiVZgwth1yjB2gwAKIIMAwFtCCC8E40PpAA=).

EgorBo replied:
Yeah, not handled even in Main 🙁 looking at it

stephentoub committed:
Streamline a few members of char

- Remove a branch from IsLetterOrDigit
- Slightly shrink the asm for IsPunctuation and IsSymbol
- Remove a bounds check from IsSurrogatePair
- Remove some bounds checks, other branching, and duplicative code from ConvertToUtf32
